quickwit-search/src/search_stream/leaf.rs

```diff
@@ -46,39 +55,52 @@ pub async fn leaf_search_stream(
     storage: Arc<dyn Storage>,
     splits: Vec<SplitIdAndFooterOffsets>,
     index_config: Arc<dyn IndexConfig>,
-) -> UnboundedReceiverStream<Result<LeafSearchStreamResult, tonic::Status>> {
+) -> UnboundedReceiverStream<crate::Result<LeafSearchStreamResult>> {
     let (result_sender, result_receiver) = tokio::sync::mpsc::unbounded_channel();
-    for split in splits {
-        let index_config_clone = index_config.clone();
-        let result_sender_clone = result_sender.clone();
-        let request_clone = request.clone();
-        let storage_clone = storage.clone();
-        tokio::spawn(async move {
-            let leaf_split_result = leaf_search_stream_single_split(
-                split.clone(),
-                index_config_clone,
-                &request_clone,
-                storage_clone,
-            )
-            .await
-            .map_err(SearchError::convert_to_tonic_status);
-            result_sender_clone.send(leaf_split_result).map_err(|_| {
-                SearchError::InternalError(format!(
-                    "Unable to send leaf export result for split `{}`",
-                    &split.split_id
-                ))
-            })?;
-            Result::<(), SearchError>::Ok(())
-        });
-    }
+    let request_clone = request.clone();
+    tokio::spawn(async move {
+        let mut stream =
+            leaf_search_results_stream(request_clone, storage, splits, index_config).await;
+        while let Some(item) = stream.next().await {
+            if let Err(error) = result_sender.send(item) {
+                error!(
+                    "Failed to send leaf search stream result. Stop sending. Cause: {}",
+                    error
+                );
+                break;
+            }
+        }
+    });
     UnboundedReceiverStream::new(result_receiver)
 }
 
+async fn leaf_search_results_stream(
+    request: SearchStreamRequest,
+    storage: Arc<dyn Storage>,
+    splits: Vec<SplitIdAndFooterOffsets>,
+    index_config: Arc<dyn IndexConfig>,
+) -> impl futures::Stream<Item = crate::Result<LeafSearchStreamResult>> + Sync + Send + 'static {
+    futures::stream::iter(splits)
+        .map(move |split| {
+            (
+                split,
+                index_config.clone(),
+                request.clone(),
+                storage.clone(),
+            )
+        })
+        .map(|(split, index_config_clone, request_clone, storage)| {
```

Comment:
Why the double `map`? Readability?

Reply:
readability and compiler issue I had but it's no more relevant.